vernac/egramcoq.ml

```diff
@@ -434,8 +434,10 @@ type ('self, _, 'r) ty_rule =
 
 type 'r gen_eval = Loc.t -> 'r env -> 'r
 
+let outer_eval n a b c =
 let rec ty_eval : type s a. (s, a, Loc.t -> s) ty_rule -> s gen_eval -> s env -> a = function
 | TyStop ->
+  Stats.parser_action "NOTATION" n "??" 999;
```

Comment:
@herbelin I'm having trouble getting this code to work correctly for a few notations.  Perhaps you can help me understand what I need to fix?  (If you're still busy with 8.13, I can wait a while).

The code builds parse trees and verifies that every input sentence leaves a single entry in my parse stack.  If there's more than one entry at the end of a sentence or a stack underflow, that means the parse tree is broken.  In which case, the code does an `exit 98`.  I can generate parse trees for everything in the source tree and all but 5 of the CI tasks.  I don't understand why it fails in a few cases, but I think it's related to a corner case of notations.

`Stats.parser_action` indicates that a production has been recognized.  The parameters are: a production identifier, the number of items in the production and the file and line number where the production was defined.  (Currently I don't have code to pass the correct file and line number.)

In ci-iris, I believe this input:

```
Goal forall P1 P2 P3 P4 (P: Prop), P.
```

uses this notation defined in ascii.v:

```
Notation "'forall' x .. y , P" :=
  (bi_forall (fun x => .. (bi_forall (fun y => P%I)) ..))
  (at level 200, x binder, right associativity, only parsing) : bi_scope.
```

In which case, the second `Reduce` in the following excerpt should not happen--right?  That reduction is for the production `binder_constr -> "forall" open_binders "," term200`, exactly what the notation accepts (the third reduction).  I hope you can steer me in the right direction.

(The top of the stack is printed last.  `[ ... ]` surrounds productions, lists, etc.  Tokens are printed as-is.  The output is fairly readable but not polished.)

```
Reduce term0-1 1 parsing/g_constr.mlg 178
  > [Slist0 ]
  > [Slist0 ]
  > Goal
  > forall
  > [open_binders-2 [name-2 [ident-1 P1 ]][Slist0 [name-2 [ident-1 P2 ]][name-2 [ident-1 P3 ]][name-2 [ident-1 P4 ]]][binders-1 [Slist0 [binder-2 [closed_binder-2 ( [name-2 [ident-1 P ]]: [lconstr-1 [term0-1 [atomic_constr-2 [sort-2 Prop ]]]]) ]]]]]
  > ,
  > [term0-1 [atomic_constr-1 [global-1 [reference-2 [ident-1 P ]]][univ_annot-2 ]]]

Reduce binder_constr-1 4 parsing/g_constr.mlg 217
  > [Slist0 ]
  > [Slist0 ]
  > Goal
  > [binder_constr-1 forall [open_binders-2 [name-2 [ident-1 P1 ]][Slist0 [name-2 [ident-1 P2 ]][name-2 [ident-1 P3 ]][name-2 [ident-1 P4 ]]][binders-1 [Slist0 [binder-2 [closed_binder-2 ( [name-2 [ident-1 P ]]: [lconstr-1 [term0-1 [atomic_constr-2 [sort-2 Prop ]]]]) ]]]]], [term0-1 [atomic_constr-1 [global-1 [reference-2 [ident-1 P ]]][univ_annot-2 ]]]]

Reduce NOTATION 4 ?? 999
  > [NOTATION [Slist0 ][Slist0 ]Goal [binder_constr-1 forall [open_binders-2 [name-2 [ident-1 P1 ]][Slist0 [name-2 [ident-1 P2 ]][name-2 [ident-1 P3 ]][name-2 [ident-1 P4 ]]][binders-1 [Slist0 [binder-2 [closed_binder-2 ( [name-2 [ident-1 P ]]: [lconstr-1 [term0-1 [atomic_constr-2 [sort-2 Prop ]]]]) ]]]]], [term0-1 [atomic_constr-1 [global-1 [reference-2 [ident-1 P ]]][univ_annot-2 ]]]]]
```